treewide: Add passthru.updateScript to gnome pkgs

JohnAZoidberg opened the pull request:
###### Motivation for this change
For packages that are hosted on the gnome mirror it's very easy.
Inspired by https://github.com/NixOS/nixpkgs/issues/36150

I changed some source URLs but they should be equivalent to the original.

Changed some derivations to use `pname` instead of `name` so that causes many package rebuilds :/

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->



###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

JohnAZoidberg committed:
treewide: Add passthru.updateScript to gnome pkgs

`updateScript` lets us update packages automatically.
For packages that are hosted on the gnome mirror it's very easy to add
that. Inspired by https://github.com/NixOS/nixpkgs/issues/36150

JohnAZoidberg commented:
Oh I should base this off of staging right?
Or staging-next? Which branch was it?

jtojnar commented on pkgs/development/tools/documentation/gtk-doc/default.nix:
```diff
@@ -38,6 +38,9 @@ stdenv.mkDerivation rec {
     # Consumers are expected to copy the m4 files to their source tree, let them reuse the patch
     respect_xml_catalog_files_var_patch = ./respect-xml-catalog-files-var.patch;
+    updateScript = gnome3.updateScript {
+      packageName = pname;
```
```suggestion
      packageName = pname;
      versionPolicy = "none";
```

jtojnar commented on pkgs/tools/security/polkit-gnome/default.nix:
```diff
@@ -20,7 +17,14 @@ in stdenv.mkDerivation rec {
+  passthru = {
+    updateScript = gnome3.updateScript {
+      packageName = pname;
+      attrPath = "polkit_gnome";
```
```suggestion
      attrPath = "polkit_gnome";
      versionPolicy = "none";
```

jtojnar commented on pkgs/development/libraries/libnotify/default.nix:
```diff
@@ -17,6 +16,12 @@ stdenv.mkDerivation rec {
 
+  passthru = {
+    updateScript = gnome3.updateScript {
+      packageName = pname;
```
```suggestion
      packageName = pname;
      versionPolicy = "none";
```

jtojnar commented on pkgs/development/libraries/hyena/default.nix:
```diff
@@ -25,6 +25,12 @@ stdenv.mkDerivation rec {
 
+  passthru = {
+    updateScript = gnome3.updateScript {
+      packageName = pname;
```
```suggestion
      packageName = pname;
      versionPolicy = "none";
```

jtojnar commented on pkgs/development/compilers/vala/default.nix:
```diff
@@ -91,6 +91,12 @@ let
     doCheck = false; # fails, requires dbus daemon
 
+    passthru = {
+      updateScript = gnome3.updateScript {
```
Update script unfortunately does not work for Vala, due to the specific way the derivation is structured. We would need to update it like this:

```patch
--- a/pkgs/development/compilers/vala/default.nix
+++ b/pkgs/development/compilers/vala/default.nix
@@ -1,17 +1,15 @@
 { stdenv, lib, fetchurl, fetchpatch, pkgconfig, flex, bison, libxslt, autoconf, automake, autoreconfHook
-, graphviz, glib, libiconv, libintl, libtool, expat, substituteAll
+, graphviz, glib, libiconv, libintl, libtool, expat, substituteAll, gnome3
 }:
 
 let
   generic = lib.makeOverridable ({
-    major, minor, sha256,
+    version, sha256,
     extraNativeBuildInputs ? [],
     extraBuildInputs ? [],
     withGraphviz ? false
   }:
   let
-    atLeast = lib.versionAtLeast "${major}.${minor}";
-
     # Patches from the openembedded-core project to build vala without graphviz
     # support. We need to apply an additional patch to allow building when the
     # header file isn't available at all, but that patch (./gvc-compat.patch)
@@ -45,21 +43,21 @@
         #     0.42.4: https://github.com/openembedded/openembedded-core/raw/f2b4f9ec6f44dced7f88df849cca68961419eeb8/meta/recipes-devtools/vala/vala/disable-graphviz.patch
         "0.44" = ./disable-graphviz-0.44.1.patch;
 
-      }.${major} or (throw "no graphviz patch for this version of vala");
+      }.${lib.versions.majorMinor version} or (throw "no graphviz patch for this version of vala");
 
-    disableGraphviz = atLeast "0.38" && !withGraphviz;
+    disableGraphviz = lib.versionAtLeast version "0.38" && !withGraphviz;
 
   in stdenv.mkDerivation rec {
-    name = "vala-${version}";
-    version = "${major}.${minor}";
+    pname = "vala";
+    inherit version;
 
     setupHook = substituteAll {
       src = ./setup-hook.sh;
-      apiVersion = major;
+      apiVersion = lib.versions.majorMinor version;
     };
 
     src = fetchurl {
-      url = "mirror://gnome/sources/vala/${major}/${name}.tar.xz";
+      url = "mirror://gnome/sources/${pname}/${lib.versions.majorMinor version}/${pname}-${version}.tar.xz";
       inherit sha256;
     };
 
@@ -78,19 +76,26 @@
 
     nativeBuildInputs = [
       pkgconfig flex bison libxslt
-    ] ++ lib.optional (stdenv.isDarwin && (atLeast "0.38")) expat
+    ] ++ lib.optional (stdenv.isDarwin && (lib.versionAtLeast version "0.38")) expat
       ++ lib.optional disableGraphviz autoreconfHook # if we changed our ./configure script, need to reconfigure
       ++ extraNativeBuildInputs;
 
     buildInputs = [
       glib libiconv libintl
-    ] ++ lib.optional (atLeast "0.38" && withGraphviz) graphviz
+    ] ++ lib.optional (lib.versionAtLeast version "0.38" && withGraphviz) graphviz
       ++ extraBuildInputs;
 
     enableParallelBuilding = true;
 
     doCheck = false; # fails, requires dbus daemon
 
+    # passthru = {
+    #   updateScript = gnome3.updateScript {
+    #     packageName = pname;
+    #     attrPath = "${pname}_${lib.versions.major version}_${lib.versions.minor version}";
+    #   };
+    # };
+
     meta = with stdenv.lib; {
       description = "Compiler for GObject type system";
       homepage = https://wiki.gnome.org/Projects/Vala;
@@ -102,33 +107,28 @@
 
 in rec {
   vala_0_36 = generic {
-    major   = "0.36";
-    minor   = "18";
+    version = "0.36.18";
     sha256  = "0csb9skgy663y05kl813dsarzjbfcdsmx5fvz13p8gas3hycciq9";
   };
 
   vala_0_38 = generic {
-    major   = "0.38";
-    minor   = "10";
+    version = "0.38.10";
     sha256  = "1rdwwqs973qv225v8b5izcgwvqn56jxgr4pa3wxxbliar3aww5sw";
     extraNativeBuildInputs = [ autoconf ] ++ lib.optional stdenv.isDarwin libtool;
   };
 
   vala_0_40 = generic {
-    major   = "0.40";
-    minor   = "14";
+    version = "0.40.14";
     sha256  = "0llid9b9cgjcrcclc0pw2skkssb7br7b2clq9cql3p14dl94gki0";
   };
 
   vala_0_42 = generic {
-    major   = "0.42";
-    minor   = "6";
+    version = "0.42.6";
     sha256  = "14024gvs23q323fmd62hqd8jiypaxbjjvamyd782ixbhxmpz8x1p";
   };
 
   vala_0_44 = generic {
-    major   = "0.44";
-    minor   = "1";
+    version = "0.44.1";
     sha256 = "1qiglkgymws6a3m8xz1v8b3na165ywx2dbipp3bdg134bi0w3a4n";
   };
 
```

But even then there are some more problems, mainly the fact that the `updateScript`s are run in parallel. I suggest you use the patch above but not include the `updateScript` itself until #59372 is merged.

JohnAZoidberg replied:
Yeah haha, I had my suspicions that it wouldn't work. Thanks for the patch, I just had to fix some small things!

jtojnar commented:
Thanks for taking care of this!

----

GNOME update script ignores the versions with odd minors by default because they are usually considered unstable. For projects where it is not the case, or we want the unstable version, we need to override the `versionPolicy` attribute.

Yes, `staging` is the correct branch.

JohnAZoidberg commented:
How did you determine for which packages this is not the case?
Those that are in nixpkgs with an odd minor?

jtojnar commented:
> How did you determine for which packages this is not the case?
> Those that are in nixpkgs with an odd minor?

Yup, just checked the minor part of version. Maybe some of those are unstable by accident, we should probably check the repos.

JohnAZoidberg commented:
I looked at the repos and it doesn't seem that any of them use odd versions to indicate unstable. I'm not actually sure it's so common that it should be the default policy for us.

However I noticed that some of the packages have been abandoned for quite a while and aren't used anymore - I'll probably remove them in another PR or start a discussion in an issue.

jtojnar commented:
> I looked at the repos and it doesn't seem that any of them use odd versions to indicate unstable. I'm not actually sure it's so common that it should be the default policy for us.

I would say about 90% of packages hosted on GNOME servers, the packages with `updateScript` should have a correct version policy set so `grep`ing nixpkgs can give you pretty acurate image.

> However I noticed that some of the packages have been abandoned for quite a while and aren't used anymore - I'll probably remove them in another PR or start a discussion in an issue.

At least glabels and ekiga have a new major release in the works. opal will probably be updated with ekiga.

Maybe we can get rid of pangox-compat.

JohnAZoidberg commented:
> > I looked at the repos and it doesn't seem that any of them use odd versions to indicate unstable. I'm not actually sure it's so common that it should be the default policy for us.
> 
> I would say about 90% of packages hosted on GNOME servers, the packages with `updateScript` should have a correct version policy set so `grep`ing nixpkgs can give you pretty acurate image.

What do you mean?
Do we need to add the policy? Do they have that policy upstream?
Ohh okay, this odd policy is just the default for GNOME packages, alright.
I take my statement back, it's probably the right default.

> > However I noticed that some of the packages have been abandoned for quite a while and aren't used anymore - I'll probably remove them in another PR or start a discussion in an issue.
> 
> At least glabels and ekiga have a new major release in the works. opal will probably be updated with ekiga.
> 
> Maybe we can get rid of pangox-compat if nothing else depends on it.
I'm speaking about hyena (and the only program that uses it: pdf-mod) and polkit-gnome.

JohnAZoidberg commented:
I'll add some more tonight.

jtojnar commented:
Do we actually want to add update scripts to packages that will never receive an update? In particular those under Archive namespace like https://gitlab.gnome.org/Archive/pygtk

worldofpeace commented on pkgs/data/fonts/ttf-bitstream-vera/default.nix:
```diff
@@ -1,9 +1,9 @@
   name = "ttf-bitstream-vera-1.10";
 
-  url = mirror://gnome/sources/ttf-bitstream-vera/1.10/ttf-bitstream-vera-1.10.tar.bz2;
+  url = "mirror://gnome/sources/ttf-bitstream-vera/${stdenv.lib.versions.majorMinor version}/${name}.tar.bz2";
```
Eval error here because `version` does not exist.

You'll need to use `pname` and `version` above.

JohnAZoidberg replied:
yeah, i've got a fix on my local tree.

JohnAZoidberg commented:
Not really no. I'd like to mark them as "non maintained" or "abandoned" somehow, so that it's obvious in nixpkgs.
I think this would be useful in general. Then other general update scripts could ignore those packages.
What would you think if I added a message to `meta.knownVulnerabilities`?

JohnAZoidberg commented:
> Not really, no. I'd like to mark them as "non maintained" or "abandoned" somehow, so that it's obvious in nixpkgs.
> I think this would be useful in general. Then other general update scripts could ignore those packages.
> What would you think if I added a message to `meta.knownVulnerabilities`?

Well `meta.knownVulnerabilities` is probably not a good idea since that makes the package opt-in - as ofborg shows.

worldofpeace commented:
@JohnAZoidberg I merged https://github.com/NixOS/nixpkgs/pull/60198/commits/393ca4cf70081b09168083cb7acc5b9548835582 in #60548 so you can drop it here.

JohnAZoidberg committed:
treewide: Update meta of packages hosted on gnome

- Update homepage URLs
- Add missing meta entries

JohnAZoidberg commented:
Alright, I went through each of them and removed the `updateScript` for abandoned packages.
While I was at it I also corrected the URLs for their hompages and fixed some other meta attributes.

jtojnar commented on pkgs/applications/graphics/gthumb/default.nix:
```diff
@@ -5,13 +5,12 @@
   flex, wrapGAppsHook, shared-mime-info }:
 
 let
```
Unnecessary `let`

jtojnar commented on pkgs/applications/graphics/gthumb/default.nix:
```diff
@@ -36,11 +35,12 @@ in stdenv.mkDerivation rec {
   passthru = {
     updateScript = gnome3.updateScript {
       packageName = pname;
+      versionPolicy = "none";
```
This seems to follow GNOME versioning policy:

https://mail.gnome.org/archives/gthumb-list/2019-February/msg00000.html

JohnAZoidberg replied:
wow, good catch! I had only looked here: https://gitlab.gnome.org/GNOME/gthumb/blob/master/NEWS

jtojnar replied:
Yeah, they often assume everyone is familiar with the versioning schema and then there are exceptions :confused:

jtojnar commented on pkgs/development/libraries/gcab/default.nix:
```diff
@@ -22,10 +22,17 @@ stdenv.mkDerivation rec {
   meta = with stdenv.lib; {
     platforms = platforms.linux;
     license = licenses.lgpl21;
+    homepage = "https://github.com/GNOME/gcab";
```
https://wiki.gnome.org/msitools

jtojnar commented on pkgs/development/libraries/gstreamer/gstreamermm/default.nix:
```diff
@@ -20,9 +16,16 @@ stdenv.mkDerivation rec {
+  passthru = {
+    updateScript = gnome3.updateScript {
+      packageName = pname;
+      versionPolicy = "none";
```
This most definitely uses GNOME version policy, as both GStreamer and GLibmm do.

JohnAZoidberg replied:
I saw that they marked a lot of even versions as unstable.
https://gitlab.gnome.org/GNOME/gstreamermm/blob/master/NEWS

jtojnar replied:
Wow. Could you add a comment like “unpredictable version stability” then?

jtojnar commented on pkgs/development/libraries/libwnck/default.nix:
```diff
@@ -21,8 +18,16 @@ stdenv.mkDerivation rec {
+  passthru = {
+    updateScript = gnome3.updateScript {
+      packageName = pname;
+    };
```
This is actually an unmaintained version branch of `libwnck` so `updateScript` is pointless here. Will need to carry out https://github.com/NixOS/nixpkgs/issues/40015, but that is a separate issue.

JohnAZoidberg replied:
Oh oops. I tried to avoid updating old version, but here the old file is called `default.nix`.

jtojnar commented on pkgs/development/python-modules/pygtk/default.nix:
```diff
@@ -50,6 +48,20 @@ buildPythonPackage rec {
   '';
+
+  passthru = {
+    updateScript = gnome3.updateScript {
```
pygtk is archived as well:

https://gitlab.gnome.org/Archive/pygtk

jtojnar commented on pkgs/development/python-modules/pygtk/default.nix:
```diff
@@ -50,6 +48,20 @@ buildPythonPackage rec {
+  };
+
+  meta = with stdenv.lib; {
+    description = "Python package which provides bindings for GObject based libraries such as GTK, GStreamer, WebKitGTK, GLib, GIO and many more";
```
pygobject is a successor project, pygtk only provided bindings for GTK 2.

jtojnar commented on pkgs/development/tools/documentation/gnome-doc-utils/default.nix:
```diff
@@ -20,4 +22,17 @@ python2Packages.buildPythonApplication {
+  };
+
+  meta = with stdenv.lib; {
+    description = "Collection of documentation utilities for the Gnome project";
```
```suggestion
    description = "Collection of documentation utilities for the GNOME project";
```

jtojnar commented on pkgs/misc/themes/gtk2/gtk-engine-murrine/default.nix:
```diff
@@ -1,16 +1,23 @@
   buildInputs = [ intltool gtk2 ];
 
+  passthru = {
+    updateScript = gnome3.updateScript {
```
Theme engines are GTK2 only feature and has been archived: https://gitlab.gnome.org/Archive/murrine

JohnAZoidberg replied:
I read that, too. But I saw that Ubuntu has them for GTK3 (or it's just bad naming): https://packages.ubuntu.com/search?keywords=gtk3-engines

I'll remove the `updateScript`

jtojnar replied:
Hmm, apparently the support was not removed https://blogs.gnome.org/mclasen/2014/06/13/a-new-default-theme-for-gtk/, they are just frowned upon now.

JohnAZoidberg replied:
Interesting!

jtojnar commented on pkgs/tools/security/polkit-gnome/default.nix:
```diff
@@ -20,10 +17,10 @@ in stdenv.mkDerivation rec {
-    homepage = https://hal.freedesktop.org/docs/PolicyKit/;
+    homepage = "https://gitlab.gnome.org/Archive/policykit-gnome";
     description = "A dbus session bus service that is used to bring up authentication dialogs";
     license = stdenv.lib.licenses.gpl2;
```
```suggestion
    license = stdenv.lib.licenses.lgpl2Plus;
```

https://gitlab.gnome.org/Archive/policykit-gnome/blob/a0763a246a81188f60b0f9810143e49224dc752f/src/main.c#L5-7

Unfortunately, some of the older packages were not checked very rigorously.

jtojnar commented on pkgs/development/libraries/ptlib/default.nix:
```diff
@@ -37,6 +38,7 @@ stdenv.mkDerivation rec {
   meta = with stdenv.lib; {
     description = "Portable Tools from OPAL VoIP";
     maintainers = [ maintainers.raskin ];
+    homepage = "https://github.com/PeteDavidson/ptlib";
```
```suggestion
    homepage = "http://www.opalvoip.org/";
```

jtojnar commented on pkgs/development/python-modules/pygtk/default.nix:
```diff
@@ -50,6 +48,20 @@ buildPythonPackage rec {
+                  $out/lib/${python.libPrefix}/site-packages/${pname}-${version}.pth
   '';
+
+  meta = with stdenv.lib; {
```
Am I seeing double :laughing:

FRidh commented:
What's the status?

worldofpeace commented:
I think it should be fine, (haven't really looked at everything it's a lot) just need to resolve conflicts.

JohnAZoidberg commented:
Alright, I rebased. Just needed to fix that one tiny conflict.
Didn't change anything else ;)

worldofpeace commented on pkgs/development/libraries/libxmlxx/default.nix:
```diff
@@ -16,11 +15,18 @@ stdenv.mkDerivation rec {
     license = licenses.lgpl2Plus;
     platforms = platforms.unix;
     maintainers = with maintainers; [ phreedom ];
+    repositories.git = "https://gitlab.gnome.org/GNOME/libxmlplusplus/";
```
```suggestion
```

Think we're good for merge after that

JohnAZoidberg replied:
We've got `meta.repositories.git` in the tree 74 times.
I think if the repo is neither at `src.url` or `meta.homepage` it's okay to have.

worldofpeace replied:
see https://github.com/NixOS/nixpkgs/issues/61435. if ratified perhaps we can use it widespread.

JohnAZoidberg replied:
Alright, thanks for creating the issue.

worldofpeace commented:
Removed the reference to `repositories.git`.

worldofpeace reviewed (approved):
Excluding the things in `pkgs/desktops/gnome-2` we've gone from `40` to `24` occurrences that
the command to track this in https://github.com/NixOS/nixpkgs/issues/36150 can catch.

worldofpeace commented:
Thanks @JohnAZoidberg for the fixes :sparkles:
